src/device/array.jl

```diff
@@ -0,0 +1,155 @@
+
+# NOTE: we can't support the typical `tuple or series of integer` style construction,
+#       because we're currently requiring a trailing pointer argument.
+
+struct VeDeviceArray{T,N,A} <: AbstractArray{T,N}
+    shape::Dims{N}
+    ptr::LLVMPtr{T,A}
+
+    # inner constructors, fully parameterized, exact types (ie. Int not <:Integer)
+    VeDeviceArray{T,N,A}(shape::Dims{N}, ptr::LLVMPtr{T,A}) where {T,A,N} = new(shape,ptr)
+end
+
+const VeDeviceVector = VeDeviceArray{T,1,A} where {T,A}
+const VeDeviceMatrix = VeDeviceArray{T,2,A} where {T,A}
+
+# outer constructors, non-parameterized
+VeDeviceArray(dims::NTuple{N,<:Integer}, p::LLVMPtr{T,A})                where {T,A,N} = VeDeviceArray{T,N,A}(dims, p)
+VeDeviceArray(len::Integer,              p::LLVMPtr{T,A})                where {T,A}   = VeDeviceVector{T,A}((len,), p)
+
+# outer constructors, partially parameterized
+VeDeviceArray{T}(dims::NTuple{N,<:Integer},   p::LLVMPtr{T,A}) where {T,A,N} = VeDeviceArray{T,N,A}(dims, p)
+VeDeviceArray{T}(len::Integer,                p::LLVMPtr{T,A}) where {T,A}   = VeDeviceVector{T,A}((len,), p)
+VeDeviceArray{T,N}(dims::NTuple{N,<:Integer}, p::LLVMPtr{T,A}) where {T,A,N} = VeDeviceArray{T,N,A}(dims, p)
+VeDeviceVector{T}(len::Integer,               p::LLVMPtr{T,A}) where {T,A}   = VeDeviceVector{T,A}((len,), p)
+
+# outer constructors, fully parameterized
+VeDeviceArray{T,N,A}(dims::NTuple{N,<:Integer}, p::LLVMPtr{T,A}) where {T,A,N} = VeDeviceArray{T,N,A}(Int.(dims), p)
+VeDeviceVector{T,A}(len::Integer,               p::LLVMPtr{T,A}) where {T,A}   = VeDeviceVector{T,A}((Int(len),), p)
+
+
+## getters
+
+Base.pointer(a::VeDeviceArray) = a.ptr
+Base.pointer(a::VeDeviceArray, i::Integer) =
+    pointer(a) + (i - 1) * Base.elsize(a)
+
+Base.elsize(::Type{<:VeDeviceArray{T}}) where {T} = sizeof(T)
+Base.size(g::VeDeviceArray) = g.shape
+Base.length(g::VeDeviceArray) = prod(g.shape)
+
+
+## conversions
+
+Base.unsafe_convert(::Type{LLVMPtr{T,A}}, a::VeDeviceArray{T,N,A}) where {T,A,N} = pointer(a)
+
+
+## indexing intrinsics
+
+# NOTE: these intrinsics are now implemented using plain and simple pointer operations;
+#       when adding support for isbits union arrays we will need to implement that here.
+
+# TODO: arrays as allocated by the VEDA APIs are 256-byte aligned. we should keep track of
+#       this information, because it enables optimizations like Load Store Vectorization
+#       (cfr. shared memory and its wider-than-datatype alignment)
+
+@inline function arrayref(A::VeDeviceArray{T}, index::Int) where {T}
+    @boundscheck checkbounds(A, index)
+    align = Base.datatype_alignment(T)
+    unsafe_load(pointer(A), index, Val(align))
+end
+
+@inline function arrayset(A::VeDeviceArray{T}, x::T, index::Int) where {T}
+    @boundscheck checkbounds(A, index)
+    align = Base.datatype_alignment(T)
+    unsafe_store!(pointer(A), x, index, Val(align))
+    return A
+end
+
+@inline function const_arrayref(A::VeDeviceArray{T}, index::Int) where {T}
```

Comment:
Suspect that this won't work yet.

Reply:
I removed this const stuff to keep things simple.